src/table/boot.rs

```diff
@@ -128,17 +128,32 @@ pub struct BootServices {
     check_event: unsafe extern "efiapi" fn(event: Event) -> Status,
 
     // Protocol handlers
-    install_protocol_interface: usize,
-    reinstall_protocol_interface: usize,
-    uninstall_protocol_interface: usize,
+    install_protocol_interface: unsafe extern "efiapi" fn(
+        handle: NonNull<Option<Handle>>,
+        guid: &Guid,
+        interface_type: InterfaceType,
+        interface: Option<NonNull<c_void>>,
+    ) -> Status,
+    reinstall_protocol_interface: unsafe extern "efiapi" fn(
+        handle: Handle,
+        protocol: &Guid,
+        old_interface: Option<NonNull<c_void>>,
+        new_interface: Option<NonNull<c_void>>,
```

Comment:
`Option<NonNull<c_void>>` feels a littel redundent, I think these could just be `*mut c_void`?

Reply:
They could. My thinking was that passing a `None` was more idiomatic than a `ptr::null()`, though the `Some(e)` case is a bit messy still.

What would you think about defining a newtype similar to `Handle` and `Event`? Or possibly just using `Handle`? These are only `Handle`s to a protocol implementation struct and I believe are the same "kind" of `Handle` that would be returned from `get_handle_for_protocol` and similar.

Reply:
I think that [re]installing protocols is inherently a pretty low-level interface, so I think sticking with something closer to the C API might be preferable here to make the usage clear. I suspect passing `None`/`null` for the interface would be pretty rare anyway, since most useful protocols are going to have some kind of struct associated with them.

Reply:
That's fair, I don't actually think there are any protocols that don't have some sort of interface attached to them in edk2. I'll make the change.